### ISSUE TITLE
Restarting the same stream

### ISSUE THREAD

**Issue by mr-j0nes** [RaftLib/RaftLib#98]
Hi Jonathan,

very quick question, is there any particular reason why a stream cannot be restarted without initializing the map ? E.g.:

```
    raft::map m;
    m += a >> b >> c;
    m.exe();
    m.exe();  // <-- throws exception
```
The exception thrown is:

```
terminate called after throwing an instance of 'PortExceptionBase<1>'
  what():  Source port "out" already initialized!
```

**Comment by jonathan-beard** [comment]
## Short answer

I wanted to throw an exception if people try to attach to the same port twice and also have an exception for when the runtime itself errs and tries to double allocate things (as a sanity  check). It doesn't have to throw an exception at each exe(), it could be smart enough not to throw an exception from there (the exe) but only from the >> operators. 

## Longer answer

The graph is built lazily, and by that I mean that nothing is really evaluated aside from connectivity and type until you call exe. So what the exception is saying is that the runtime is trying to allocate another FIFO for something that already has one. Again, I think you're right @mr-j0nes, we should allow the exe to be called again. Shouldn't be too hard to fix up.

**Comment by mr-j0nes** [comment]
Exactly. Now I am trying to implement a recovering logic which requires the stream to be restarted. It does not feel right to have to re-link all kernels while they are already linked and nothing has to change in that area.

Good that you also agree, it seems to me that all checks and evaluations should be done at the kernel linking stage rather than in the exe().

Perfect, then  I wait until you have a bit more of time.

Cheers..

**Comment by jonathan-beard** [comment]
HI @mr-j0nes, I'm working on a fix, right now I'm gonna have a check in the ```exe``` for linking just as before, but there will be a bool to ensure it's not done twice. Seems like a good interim fix. Almost passing test cases, but, it seems there's a stack vs. heap alloc issue.

**Comment by mr-j0nes** [comment]
Hi Jonathan, hope you are doing fine. 
I just remember this change and wanted to know if there is something there that I can test ? Is the feature already implemented ?

**Comment by jonathan-beard** [comment]
It's in progress. There's shutdown condition that I need to deal with. More specifically, There needs to be a cleaner way of doing "barriers" to wait on the specific kernel that is needed within the main thread of execution.

**Comment by mr-j0nes** [comment]
I know there is a [branch](https://github.com/RaftLib/RaftLib/tree/restartfix) with work on this, but I have work around'ed this by just creating all kernels again in the second run:

```
void run()
{
  A a; B b; C c;
  raft::map m;
  m += a >> b >> c;
  m.exe();
}

in main()
{
  run();
  run(); // <-- second run successfully
}
```
